src/Swarm/TUI/Launch/Controller.hs

```diff
@@ -0,0 +1,155 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Event handling for the scenario launch configuration dialog.
+module Swarm.TUI.Launch.Controller where
+
+import Brick hiding (Direction, Location)
+import Brick.Focus
+import Brick.Widgets.Edit (handleEditorEvent)
+import Brick.Widgets.FileBrowser
+import Brick.Widgets.FileBrowser qualified as FB
+import Control.Lens
+import Control.Monad.Except (forM_, liftIO, when)
+import Data.Maybe (listToMaybe)
+import Graphics.Vty qualified as V
+import Swarm.Game.ScenarioInfo
+import Swarm.TUI.Controller.Util
+import Swarm.TUI.Launch.Model
+import Swarm.TUI.Launch.Prep (initFileBrowserWidget, makeFocusRingWith, parseWidgetParams, toValidatedParams)
+import Swarm.TUI.Model
+import Swarm.TUI.Model.Name
+import Swarm.TUI.Model.StateUpdate
+import Swarm.TUI.Model.UI
+import Swarm.Util (listEnums)
+
+cacheValidatedInputs :: EventM Name LaunchOptions ()
+cacheValidatedInputs = do
+  launchControls <- use controls
+  parsedParams <- liftIO $ parseWidgetParams launchControls
```

Comment:
> Typing in the initial seed field takes a long time (i.e. it takes a second or two to update after every keypress), I wonder why?

Here's why: the seed field and the script field are currently getting validated together upon every dialog event.  The script field is validated by parsing the code.  I'll look into separating the validation of each field.